### PR TITLE
[sdpa] Fix export error with unbacked symbolic ints in attn_mask shape check

### PR DESCRIPTION
**Pull request by LuFinch**
torch.export.export() fails with GuardOnDataDependentSymNode when tracing F.scaled_dot_product_attention on XPU with a 4D attention mask whose batch dimension is an unbacked symbolic integer (e.g., from data-dependent filtering).

Both CPU/XPU has this issue. see https://github.com/intel/torch-xpu-ops/issues/3418

# Rootcause
In check_attn_mask_shape (aten/src/ATen/native/transformers/sdp_utils_cpp.h), the mask shape validation compares symbolic dimensions against literal 1 using sym_size(0) == 1. During torch.export, if the batch dimension is an unbacked symbolic int u0 (produced by data-dependent operations like boolean indexing), this comparison triggers guard_bool on Eq(u0, 1), which is not allowed for data-dependent symbols.

# Fix
Replace direct sym_size() == 1 / sym_size() != 1 comparisons with
```
auto dim_compatible = [](const c10::SymInt& mask_dim,
                           const c10::SymInt& target_dim) -> bool {
    if (TORCH_STATICALLY_KNOWN_TRUE(mask_dim == target_dim)) {
      return true;
    }
    auto mask_int = mask_dim.maybe_as_int();
    return mask_int.has_value() && *mask_int == 1;
  };
```

### CODE REVIEW

**Comment by LuFinch** [comment]
@guangyey @etaf Could you help review?

**Comment by etaf** [comment]
The commit message and title say [xpu] but this is a shared header (sdp_utils_cpp.h) that affects CPU and potentially CUDA cuDNN paths too. A more accurate title would be [sdpa] Fix export error with unbacked symbolic ints in attn_mask shape check.

**Comment by etaf** [comment]
LGTM

**Review by guangyey — COMMENTED** [review]
Need a test case!

File: aten/src/ATen/native/transformers/sdp_utils_cpp.h [hunk 1]
```diff
@@ -278,22 +278,34 @@ inline bool check_attn_mask_shape(sdp_params const& params, bool debug) {
   auto qSize = params.query.sym_size(2);
   auto kvSize = params.key.sym_size(2);
   auto num_head = params.query.sym_size(1);
-  if (attn_mask.value().sym_size(-2) != qSize && attn_mask.value().sym_size(-2) != 1) {
+
+  // Use maybe_as_int() for comparisons to avoid guarding
+  // on unbacked symbolic ints (e.g. data-dependent dims during torch.export).
+  auto mask_seq = attn_mask.value().sym_size(-2);
+  if (mask_seq != qSize &&
+      mask_seq.maybe_as_int() != std::make_optional<int64_t>(1)) {
```
Comment (guangyey) [inline aten/src/ATen/native/transformers/sdp_utils_cpp.h]: I think `TORCH_GUARD_OR_TRUE` is what you really expect! Could you please help double confirm.

**Review by jansel — CHANGES_REQUESTED** [review]
Add a test for what this is fixing

**Comment by LuFinch** [comment]
Add UTs, please help review.

File: aten/src/ATen/native/transformers/sdp_utils_cpp.h (hunk 1 above)
Comment (LuFinch) [reply aten/src/ATen/native/transformers/sdp_utils_cpp.h]: Claude help me investigate: `TORCH_GUARD_OR_TRUE` inserts a runtime guard that requires a concrete value for validation, but unbacked symbols have no concrete value to verify against, so it would error out instead of gracefully falling back to another backend. `TORCH_STATICALLY_KNOWN_TRUE` is the correct choice: it returns the result if symbolically determinable, otherwise conservatively returns false without inserting any guard, allowing the backend selection logic to safely skip incompatible backends.

Comment (guangyey) [reply aten/src/ATen/native/transformers/sdp_utils_cpp.h]: Sounds good!

File: test/test_transformers.py [hunk 2]
```diff
@@ -2229,6 +2229,40 @@ def test_sdpa_output_shape_uses_value_head_dim(self, device):
...
+                self.hidden_size = self.num_heads * self.head_dim
+                # 3 * 12 * 12 = 432
+                self.proj_q = nn.Linear(432, self.hidden_size)
+                self.proj_k = nn.Linear(432, self.hidden_size)
+                self.proj_v = nn.Linear(432, self.hidden_size)
+
+            def forward(self, x):
+                keep = x.sum(dim=(-1, -2, -3)) != 0 # x: [B, 3, 12, 12]
+                x = x[keep] # [u0, 3, 12, 12].
+                flat = x.flatten(1) # [u0, 432]
```
Comment (guangyey) [inline test/test_transformers.py]:
```suggestion
                x = x[keep]  # [u0, 3, 12, 12].
                flat = x.flatten(1)  # [u0, 432]
```

**Review by guangyey — APPROVED** [review]

**Comment by jansel** [comment]
@claude look into those CI failures

**Review by jansel — APPROVED** [review]
